src/metabase/actions/execution.clj

```diff
@@ -203,22 +203,41 @@
       (execute-custom-action! action request-parameters)
       (throw (ex-info (tru "Unknown action type {0}." (name (:type action))) action)))))
 
+(defn- execute-table-action!
+  [kind table-id request-parameters]
+  (let [args
+        {:type     :query
+         :query    {:source-table table-id}
+         :database (api/check-404 (t2/select-one-fn :db_id [:model/Table :db_id] table-id))
+         :table-id table-id
+         :arg      [request-parameters]}
+
+        opts
+        {:policy :data-editing}]
+    (actions/perform-action-with-single-input-and-output kind args opts)))
+
 (mu/defn execute-dashcard!
   "Execute the given action in the dashboard/dashcard context with the given parameters
    of shape `{<parameter-id> <value>}."
   [dashboard-id       :- ::lib.schema.id/dashboard
    dashcard-id        :- ::lib.schema.id/dashcard
    request-parameters :- [:maybe [:map-of :string :any]]]
-  (let [dashcard (api/check-404 (t2/select-one :model/DashboardCard
-                                               :id dashcard-id
-                                               :dashboard_id dashboard-id))
-        action (api/check-404 (action/select-action :id (:action_id dashcard)))]
-    (analytics/track-event! :snowplow/action
-                            {:event     :action-executed
-                             :source    :dashboard
-                             :type      (:type action)
-                             :action_id (:id action)})
-    (execute-action! action request-parameters)))
+  (let [dashcard     (api/check-404 (t2/select-one :model/DashboardCard
+                                                   :id dashcard-id
+                                                   :dashboard_id dashboard-id))
+        action-id    (:action_id dashcard)
+        table-action (-> dashcard :visualization_settings :table_action)]
+    (api/check-404 (or action-id table-action))
+    (if table-action
+      (let [{:keys [kind table_id]} table-action]
+        (execute-table-action! (keyword kind) table_id request-parameters))
+      (let [action (api/check-404 (action/select-action :id (:action_id dashcard)))]
```

Comment:
nit: We already have `action-id` in scope, don't need to get it again.

Reply:
44f7713c1d10